File: src/main/java/com/ptaf/handlers/LocatorHandler.java
package com.ptaf.handlers;

import com.microsoft.playwright.FrameLocator;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.AriaRole;

/**
 * LocatorHandler provides methods for retrieving Locators based on various locator types
 * from a given Playwright Page, FrameLocator, or another Locator for chaining.
 * It serves as a utility class for abstraction around how different types of elements
 * are located within the UI.
 * <p>
 * Usage:
 * - Use the getLocatorForType method to obtain a Locator based on its type and context.
 */
public class LocatorHandler {

    /**
     * Retrieves the locator for a specific type from a Playwright Page.
     *
     * @param locatorType The type of locator (e.g., XPATH, CSS, BUTTON, etc.).
     * @param page        The Page object where the locator will be searched.
     * @param locator     The locator string or value.
     * @return A chainable Locator object corresponding to the specified type.
     * @throws IllegalArgumentException if the locator type is unknown.
     */
    public Locator getLocatorForType(String locatorType, Page page, String locator) {
        switch (locatorType.toUpperCase()) {
            case "CSS":
            case "TAG":
            case "XPATH":
                return page.locator(locator);
            case "BUTTON":
                return page.getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName(locator));
            case "LINKTEXT":
                return page.getByRole(AriaRole.LINK, new Page.GetByRoleOptions().setName(locator));
            case "OPTION":
                return page.getByRole(AriaRole.OPTION, new Page.GetByRoleOptions().setName(locator).setExact(true));
            case "TEXTBOX":
                return page.getByRole(AriaRole.TEXTBOX, new Page.GetByRoleOptions().setName(locator));
            case "CHECKBOX":
                return page.getByRole(AriaRole.CHECKBOX, new Page.GetByRoleOptions().setName(locator));
            case "RADIOBUTTON":
                return page.getByRole(AriaRole.RADIO, new Page.GetByRoleOptions().setName(locator));
            case "DROPDOWN":
                return page.getByRole(AriaRole.COMBOBOX, new Page.GetByRoleOptions().setName(locator));
            case "IMAGE":
                return page.getByRole(AriaRole.IMG, new Page.GetByRoleOptions().setName(locator));
            case "HEADING":
                return page.getByRole(AriaRole.HEADING, new Page.GetByRoleOptions().setName(locator));
            case "TAB":
                return page.getByRole(AriaRole.TAB, new Page.GetByRoleOptions().setName(locator));
            case "LIST":
                return page.getByRole(AriaRole.LIST, new Page.GetByRoleOptions().setName(locator));
            case "LISTBOX":
                return page.getByRole(AriaRole.LISTBOX, new Page.GetByRoleOptions().setName(locator));
            case "LISTITEM":
                return page.getByRole(AriaRole.LISTITEM, new Page.GetByRoleOptions().setName(locator));
            case "TABLE":
                return page.getByRole(AriaRole.TABLE, new Page.GetByRoleOptions().setName(locator));
            case "ROW":
                return page.getByRole(AriaRole.ROW, new Page.GetByRoleOptions().setName(locator));
            case "CELL":
                return page.getByRole(AriaRole.CELL, new Page.GetByRoleOptions().setName(locator));
            case "BUTTONSUBMIT":
                return page.getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName(locator).setPressed(true));
            case "SLIDER":
                return page.getByRole(AriaRole.SLIDER, new Page.GetByRoleOptions().setName(locator));
            case "SPINBUTTON":
                return page.getByRole(AriaRole.SPINBUTTON, new Page.GetByRoleOptions().setName(locator));
            case "PROGRESSBAR":
                return page.getByRole(AriaRole.PROGRESSBAR, new Page.GetByRoleOptions().setName(locator));
            case "ALERT":
                return page.getByRole(AriaRole.ALERT, new Page.GetByRoleOptions().setName(locator));
            case "ALERTDIALOG":
                return page.getByRole(AriaRole.ALERTDIALOG, new Page.GetByRoleOptions().setName(locator));
            case "DIALOG":
                return page.getByRole(AriaRole.DIALOG, new Page.GetByRoleOptions().setName(locator));
            case "NAVIGATION":
                return page.getByRole(AriaRole.NAVIGATION, new Page.GetByRoleOptions().setName(locator));
            case "MENU":
                return page.getByRole(AriaRole.MENU, new Page.GetByRoleOptions().setName(locator));
            case "MENUITEM":
                return page.getByRole(AriaRole.MENUITEM, new Page.GetByRoleOptions().setName(locator));
            case "MENUITEMCHECKBOX":
                return page.getByRole(AriaRole.MENUITEMCHECKBOX, new Page.GetByRoleOptions().setName(locator));
            case "MENUITEMRADIO":
                return page.getByRole(AriaRole.MENUITEMRADIO, new Page.GetByRoleOptions().setName(locator));
            case "TREE":
                return page.getByRole(AriaRole.TREE, new Page.GetByRoleOptions().setName(locator));
            case "TREEITEM":
                return page.getByRole(AriaRole.TREEITEM, new Page.GetByRoleOptions().setName(locator));
            case "GRID":
                return page.getByRole(AriaRole.GRID, new Page.GetByRoleOptions().setName(locator));
            case "GRIDCELL":
                return page.getByRole(AriaRole.GRIDCELL, new Page.GetByRoleOptions().setName(locator));
            case "SEPARATOR":
                return page.getByRole(AriaRole.SEPARATOR, new Page.GetByRoleOptions().setName(locator));
            case "SWITCH":
                return page.getByRole(AriaRole.SWITCH, new Page.GetByRoleOptions().setName(locator));
            case "STATUS":
                return page.getByRole(AriaRole.STATUS, new Page.GetByRoleOptions().setName(locator));
            case "BANNER":
                return page.getByRole(AriaRole.BANNER, new Page.GetByRoleOptions().setName(locator));
            case "FOOTER":
            case "CONTENTINFO":
                return page.getByRole(AriaRole.CONTENTINFO, new Page.GetByRoleOptions().setName(locator));
            case "MAIN":
                return page.getByRole(AriaRole.MAIN, new Page.GetByRoleOptions().setName(locator));
            case "COMPLEMENTARY":
                return page.getByRole(AriaRole.COMPLEMENTARY, new Page.GetByRoleOptions().setName(locator));
            case "REGION":
                return page.getByRole(AriaRole.REGION, new Page.GetByRoleOptions().setName(locator));
            case "ARTICLE":
                return page.getByRole(AriaRole.ARTICLE, new Page.GetByRoleOptions().setName(locator));
            case "FORM":
                return page.getByRole(AriaRole.FORM, new Page.GetByRoleOptions().setName(locator));
            case "LOG":
                return page.getByRole(AriaRole.LOG, new Page.GetByRoleOptions().setName(locator));
            case "MARQUEE":
                return page.getByRole(AriaRole.MARQUEE, new Page.GetByRoleOptions().setName(locator));
            case "TIMER":
                return page.getByRole(AriaRole.TIMER, new Page.GetByRoleOptions().setName(locator));
            case "TOOLTIP":
                return page.getByRole(AriaRole.TOOLTIP, new Page.GetByRoleOptions().setName(locator));
            case "TOOLBAR":
                return page.getByRole(AriaRole.TOOLBAR, new Page.GetByRoleOptions().setName(locator));
            case "PRESENTATION":
                return page.getByRole(AriaRole.PRESENTATION, new Page.GetByRoleOptions().setName(locator));
            case "FIGURE":
                return page.getByRole(AriaRole.FIGURE, new Page.GetByRoleOptions().setName(locator));
            case "TEXT":
                return page.getByText(locator);
            case "ROLE":
                return page.getByRole(AriaRole.valueOf(locator.toUpperCase()));
            case "ALTTEXT":
                return page.getByAltText(locator);
            case "TITLE":
                return page.getByTitle(locator);
            case "PLACEHOLDER":
                return page.getByPlaceholder(locator);
            case "LABEL":
                return page.getByLabel(locator);
            case "TESTID":
                return page.getByTestId(locator);
            case "ID":
                return page.locator("#" + locator);
            case "NAME":
                return page.locator("[name='" + locator + "']");
            case "CLASS":
                return page.locator("." + locator);
            default:
                throw new IllegalArgumentException("Unknown locator type: " + locatorType);
        }
    }

    /**
     * Retrieves the locator for a specific type from a FrameLocator.
     *
     * @param locatorType The type of locator (e.g., XPATH, CSS, BUTTON, etc.).
     * @param frame       The FrameLocator object where the locator will be searched.
     * @param locator     The locator string or value.
     * @return A chainable Locator object corresponding to the specified type.
     * @throws IllegalArgumentException if the locator type is unknown.
     */
    public Locator getLocatorForType(String locatorType, FrameLocator frame, String locator) {
        switch (locatorType.toUpperCase()) {
            case "CSS":
            case "TAG":
            case "XPATH":
                return frame.locator(locator);
            case "BUTTON":
                return frame.getByRole(AriaRole.BUTTON, new FrameLocator.GetByRoleOptions().setName(locator));
            case "LINKTEXT":
                return frame.getByRole(AriaRole.LINK, new FrameLocator.GetByRoleOptions().setName(locator));
            case "OPTION":
                return frame.getByRole(AriaRole.OPTION, new FrameLocator.GetByRoleOptions().setName(locator).setExact(true));
            case "TEXTBOX":
                return frame.getByRole(AriaRole.TEXTBOX, new FrameLocator.GetByRoleOptions().setName(locator));
            case "CHECKBOX":
                return frame.getByRole(AriaRole.CHECKBOX, new FrameLocator.GetByRoleOptions().setName(locator));
            case "RADIOBUTTON":
                return frame.getByRole(AriaRole.RADIO, new FrameLocator.GetByRoleOptions().setName(locator));
            case "DROPDOWN":
                return frame.getByRole(AriaRole.COMBOBOX, new FrameLocator.GetByRoleOptions().setName(locator));
            case "IMAGE":
                return frame.getByRole(AriaRole.IMG, new FrameLocator.GetByRoleOptions().setName(locator));
            case "HEADING":
                return frame.getByRole(AriaRole.HEADING, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TAB":
                return frame.getByRole(AriaRole.TAB, new FrameLocator.GetByRoleOptions().setName(locator));
            case "LIST":
                return frame.getByRole(AriaRole.LIST, new FrameLocator.GetByRoleOptions().setName(locator));
            case "LISTBOX":
                return frame.getByRole(AriaRole.LISTBOX, new FrameLocator.GetByRoleOptions().setName(locator));
            case "LISTITEM":
                return frame.getByRole(AriaRole.LISTITEM, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TABLE":
                return frame.getByRole(AriaRole.TABLE, new FrameLocator.GetByRoleOptions().setName(locator));
            case "ROW":
                return frame.getByRole(AriaRole.ROW, new FrameLocator.GetByRoleOptions().setName(locator));
            case "CELL":
                return frame.getByRole(AriaRole.CELL, new FrameLocator.GetByRoleOptions().setName(locator));
            case "BUTTONSUBMIT":
                return frame.getByRole(AriaRole.BUTTON, new FrameLocator.GetByRoleOptions().setName(locator).setPressed(true));
            case "SLIDER":
                return frame.getByRole(AriaRole.SLIDER, new FrameLocator.GetByRoleOptions().setName(locator));
            case "SPINBUTTON":
                return frame.getByRole(AriaRole.SPINBUTTON, new FrameLocator.GetByRoleOptions().setName(locator));
            case "PROGRESSBAR":
                return frame.getByRole(AriaRole.PROGRESSBAR, new FrameLocator.GetByRoleOptions().setName(locator));
            case "ALERT":
                return frame.getByRole(AriaRole.ALERT, new FrameLocator.GetByRoleOptions().setName(locator));
            case "ALERTDIALOG":
                return frame.getByRole(AriaRole.ALERTDIALOG, new FrameLocator.GetByRoleOptions().setName(locator));
            case "DIALOG":
                return frame.getByRole(AriaRole.DIALOG, new FrameLocator.GetByRoleOptions().setName(locator));
            case "NAVIGATION":
                return frame.getByRole(AriaRole.NAVIGATION, new FrameLocator.GetByRoleOptions().setName(locator));
            case "MENU":
                return frame.getByRole(AriaRole.MENU, new FrameLocator.GetByRoleOptions().setName(locator));
            case "MENUITEM":
                return frame.getByRole(AriaRole.MENUITEM, new FrameLocator.GetByRoleOptions().setName(locator));
            case "MENUITEMCHECKBOX":
                return frame.getByRole(AriaRole.MENUITEMCHECKBOX, new FrameLocator.GetByRoleOptions().setName(locator));
            case "MENUITEMRADIO":
                return frame.getByRole(AriaRole.MENUITEMRADIO, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TREE":
                return frame.getByRole(AriaRole.TREE, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TREEITEM":
                return frame.getByRole(AriaRole.TREEITEM, new FrameLocator.GetByRoleOptions().setName(locator));
            case "GRID":
                return frame.getByRole(AriaRole.GRID, new FrameLocator.GetByRoleOptions().setName(locator));
            case "GRIDCELL":
                return frame.getByRole(AriaRole.GRIDCELL, new FrameLocator.GetByRoleOptions().setName(locator));
            case "SEPARATOR":
                return frame.getByRole(AriaRole.SEPARATOR, new FrameLocator.GetByRoleOptions().setName(locator));
            case "SWITCH":
                return frame.getByRole(AriaRole.SWITCH, new FrameLocator.GetByRoleOptions().setName(locator));
            case "STATUS":
                return frame.getByRole(AriaRole.STATUS, new FrameLocator.GetByRoleOptions().setName(locator));
            case "BANNER":
                return frame.getByRole(AriaRole.BANNER, new FrameLocator.GetByRoleOptions().setName(locator));
            case "FOOTER":
            case "CONTENTINFO":
                return frame.getByRole(AriaRole.CONTENTINFO, new FrameLocator.GetByRoleOptions().setName(locator));
            case "MAIN":
                return frame.getByRole(AriaRole.MAIN, new FrameLocator.GetByRoleOptions().setName(locator));
            case "COMPLEMENTARY":
                return frame.getByRole(AriaRole.COMPLEMENTARY, new FrameLocator.GetByRoleOptions().setName(locator));
            case "REGION":
                return frame.getByRole(AriaRole.REGION, new FrameLocator.GetByRoleOptions().setName(locator));
            case "ARTICLE":
                return frame.getByRole(AriaRole.ARTICLE, new FrameLocator.GetByRoleOptions().setName(locator));
            case "FORM":
                return frame.getByRole(AriaRole.FORM, new FrameLocator.GetByRoleOptions().setName(locator));
            case "LOG":
                return frame.getByRole(AriaRole.LOG, new FrameLocator.GetByRoleOptions().setName(locator));
            case "MARQUEE":
                return frame.getByRole(AriaRole.MARQUEE, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TIMER":
                return frame.getByRole(AriaRole.TIMER, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TOOLTIP":
                return frame.getByRole(AriaRole.TOOLTIP, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TOOLBAR":
                return frame.getByRole(AriaRole.TOOLBAR, new FrameLocator.GetByRoleOptions().setName(locator));
            case "PRESENTATION":
                return frame.getByRole(AriaRole.PRESENTATION, new FrameLocator.GetByRoleOptions().setName(locator));
            case "FIGURE":
                return frame.getByRole(AriaRole.FIGURE, new FrameLocator.GetByRoleOptions().setName(locator));
            case "TEXT":
                return frame.getByText(locator);
            case "ROLE":
                return frame.getByRole(AriaRole.valueOf(locator.toUpperCase()));
            case "ALTTEXT":
                return frame.getByAltText(locator);
            case "TITLE":
                return frame.getByTitle(locator);
            case "PLACEHOLDER":
                return frame.getByPlaceholder(locator);
            case "LABEL":
                return frame.getByLabel(locator);
            case "TESTID":
                return frame.getByTestId(locator);
            case "ID":
                return frame.locator("#" + locator);
            case "NAME":
                return frame.locator("[name='" + locator + "']");
            case "CLASS":
                return frame.locator("." + locator);
            default:
                throw new IllegalArgumentException("Unknown locator type: " + locatorType);
        }
    }

    /**
     * NEW METHOD for chaining. It finds a new locator within the context of an existing base locator.
     *
     * @param locatorType The type of the new locator to chain (e.g., "BUTTON").
     * @param baseLocator The existing locator to search within.
     * @param locator     The string value for the new locator.
     * @return A new, chained Locator object.
     */
    public Locator getLocatorForType(String locatorType, Locator baseLocator, String locator) {
        switch (locatorType.toUpperCase()) {
            case "CSS":
            case "TAG":
            case "XPATH":
                return baseLocator.locator(locator);
            case "BUTTON":
                return baseLocator.getByRole(AriaRole.BUTTON, new Locator.GetByRoleOptions().setName(locator));
            case "LINKTEXT":
                return baseLocator.getByRole(AriaRole.LINK, new Locator.GetByRoleOptions().setName(locator));
            case "OPTION":
                return baseLocator.getByRole(AriaRole.OPTION, new Locator.GetByRoleOptions().setName(locator).setExact(true));
            case "TEXTBOX":
                return baseLocator.getByRole(AriaRole.TEXTBOX, new Locator.GetByRoleOptions().setName(locator));
            case "CHECKBOX":
                return baseLocator.getByRole(AriaRole.CHECKBOX, new Locator.GetByRoleOptions().setName(locator));
            case "RADIOBUTTON":
                return baseLocator.getByRole(AriaRole.RADIO, new Locator.GetByRoleOptions().setName(locator));
            case "DROPDOWN":
                return baseLocator.getByRole(AriaRole.COMBOBOX, new Locator.GetByRoleOptions().setName(locator));
            case "IMAGE":
                return baseLocator.getByRole(AriaRole.IMG, new Locator.GetByRoleOptions().setName(locator));
            case "HEADING":
                return baseLocator.getByRole(AriaRole.HEADING, new Locator.GetByRoleOptions().setName(locator));
            case "TAB":
                return baseLocator.getByRole(AriaRole.TAB, new Locator.GetByRoleOptions().setName(locator));
            case "LIST":
                return baseLocator.getByRole(AriaRole.LIST, new Locator.GetByRoleOptions().setName(locator));
            case "LISTBOX":
                return baseLocator.getByRole(AriaRole.LISTBOX, new Locator.GetByRoleOptions().setName(locator));
            case "LISTITEM":
                return baseLocator.getByRole(AriaRole.LISTITEM, new Locator.GetByRoleOptions().setName(locator));
            case "TABLE":
                return baseLocator.getByRole(AriaRole.TABLE, new Locator.GetByRoleOptions().setName(locator));
            case "ROW":
                return baseLocator.getByRole(AriaRole.ROW, new Locator.GetByRoleOptions().setName(locator));
            case "CELL":
                return baseLocator.getByRole(AriaRole.CELL, new Locator.GetByRoleOptions().setName(locator));
            case "GRID":
                return baseLocator.getByRole(AriaRole.GRID, new Locator.GetByRoleOptions().setName(locator));
            case "GRIDCELL":
                return baseLocator.getByRole(AriaRole.GRIDCELL, new Locator.GetByRoleOptions().setName(locator));
            case "TEXT":
                return baseLocator.getByText(locator);
            case "ROLE":
                return baseLocator.getByRole(AriaRole.valueOf(locator.toUpperCase()));
            case "ALTTEXT":
                return baseLocator.getByAltText(locator);
            case "TITLE":
                return baseLocator.getByTitle(locator);
            case "PLACEHOLDER":
                return baseLocator.getByPlaceholder(locator);
            case "LABEL":
                return baseLocator.getByLabel(locator);
            case "TESTID":
                return baseLocator.getByTestId(locator);
            default:
                throw new IllegalArgumentException("Unknown chained locator type: " + locatorType);
        }
    }
}

File: src/main/ptaf/ui/locator_handler.py
"""Playwright locator resolution by PTAF locator type."""

from __future__ import annotations

from typing import Protocol, Union

from playwright.sync_api import FrameLocator, Locator, Page

LocatorContext = Union[Page, FrameLocator, Locator]


class _RoleLocatorContext(Protocol):
    def locator(self, selector: str) -> Locator: ...

    def get_by_role(self, role: str, **kwargs: object) -> Locator: ...

    def get_by_text(self, text: str, **kwargs: object) -> Locator: ...

    def get_by_alt_text(self, text: str, **kwargs: object) -> Locator: ...

    def get_by_title(self, text: str, **kwargs: object) -> Locator: ...

    def get_by_placeholder(self, text: str, **kwargs: object) -> Locator: ...

    def get_by_label(self, text: str, **kwargs: object) -> Locator: ...

    def get_by_test_id(self, test_id: str) -> Locator: ...


class LocatorHandler:
    """Resolve Playwright locators from PTAF locator types across page, frame, and chain contexts."""

    def get_locator_for_type(
        self,
        locator_type: str,
        context: LocatorContext,
        locator: str,
    ) -> Locator:
        """Resolve a locator on a page, frame, or chained base locator."""
        if isinstance(context, Page):
            chained = False
        elif isinstance(context, FrameLocator):
            chained = False
        elif isinstance(context, Locator):
            chained = True
        else:
            raise TypeError(f"Unsupported locator context: {type(context)!r}")
        return self._resolve(locator_type, context, locator, chained=chained)

    def _resolve(
        self,
        locator_type: str,
        context: LocatorContext,
        locator: str,
        *,
        chained: bool,
    ) -> Locator:
        normalized = locator_type.upper()
        role_context: _RoleLocatorContext = context  # type: ignore[assignment]

        if normalized in {"CSS", "TAG", "XPATH"}:
            return role_context.locator(locator)

        if normalized == "BUTTON":
            return role_context.get_by_role("button", name=locator)
        if normalized == "LINKTEXT":
            return role_context.get_by_role("link", name=locator)
        if normalized == "OPTION":
            return role_context.get_by_role("option", name=locator, exact=True)
        if normalized == "TEXTBOX":
            return role_context.get_by_role("textbox", name=locator)
        if normalized == "CHECKBOX":
            return role_context.get_by_role("checkbox", name=locator)
        if normalized == "RADIOBUTTON":
            return role_context.get_by_role("radio", name=locator)
        if normalized == "DROPDOWN":
            return role_context.get_by_role("combobox", name=locator)
        if normalized == "IMAGE":
            return role_context.get_by_role("img", name=locator)
        if normalized == "HEADING":
            return role_context.get_by_role("heading", name=locator)
        if normalized == "TAB":
            return role_context.get_by_role("tab", name=locator)
        if normalized == "LIST":
            return role_context.get_by_role("list", name=locator)
        if normalized == "LISTBOX":
            return role_context.get_by_role("listbox", name=locator)
        if normalized == "LISTITEM":
            return role_context.get_by_role("listitem", name=locator)
        if normalized == "TABLE":
            return role_context.get_by_role("table", name=locator)
        if normalized == "ROW":
            return role_context.get_by_role("row", name=locator)
        if normalized == "CELL":
            return role_context.get_by_role("cell", name=locator)
        if normalized == "GRID":
            return role_context.get_by_role("grid", name=locator)
        if normalized == "GRIDCELL":
            return role_context.get_by_role("gridcell", name=locator)

        if not chained:
            if normalized == "BUTTONSUBMIT":
                return role_context.get_by_role(
                    "button", name=locator, pressed=True
                )
            if normalized == "SLIDER":
                return role_context.get_by_role("slider", name=locator)
            if normalized == "SPINBUTTON":
                return role_context.get_by_role("spinbutton", name=locator)
            if normalized == "PROGRESSBAR":
                return role_context.get_by_role("progressbar", name=locator)
            if normalized == "ALERT":
                return role_context.get_by_role("alert", name=locator)
            if normalized == "ALERTDIALOG":
                return role_context.get_by_role("alertdialog", name=locator)
            if normalized == "DIALOG":
                return role_context.get_by_role("dialog", name=locator)
            if normalized == "NAVIGATION":
                return role_context.get_by_role("navigation", name=locator)
            if normalized == "MENU":
                return role_context.get_by_role("menu", name=locator)
            if normalized == "MENUITEM":
                return role_context.get_by_role("menuitem", name=locator)
            if normalized == "MENUITEMCHECKBOX":
                return role_context.get_by_role("menuitemcheckbox", name=locator)
            if normalized == "MENUITEMRADIO":
                return role_context.get_by_role("menuitemradio", name=locator)
            if normalized == "TREE":
                return role_context.get_by_role("tree", name=locator)
            if normalized == "TREEITEM":
                return role_context.get_by_role("treeitem", name=locator)
            if normalized == "SEPARATOR":
                return role_context.get_by_role("separator", name=locator)
            if normalized == "SWITCH":
                return role_context.get_by_role("switch", name=locator)
            if normalized == "STATUS":
                return role_context.get_by_role("status", name=locator)
            if normalized == "BANNER":
                return role_context.get_by_role("banner", name=locator)
            if normalized in {"FOOTER", "CONTENTINFO"}:
                return role_context.get_by_role("contentinfo", name=locator)
            if normalized == "MAIN":
                return role_context.get_by_role("main", name=locator)
            if normalized == "COMPLEMENTARY":
                return role_context.get_by_role("complementary", name=locator)
            if normalized == "REGION":
                return role_context.get_by_role("region", name=locator)
            if normalized == "ARTICLE":
                return role_context.get_by_role("article", name=locator)
            if normalized == "FORM":
                return role_context.get_by_role("form", name=locator)
            if normalized == "LOG":
                return role_context.get_by_role("log", name=locator)
            if normalized == "MARQUEE":
                return role_context.get_by_role("marquee", name=locator)
            if normalized == "TIMER":
                return role_context.get_by_role("timer", name=locator)
            if normalized == "TOOLTIP":
                return role_context.get_by_role("tooltip", name=locator)
            if normalized == "TOOLBAR":
                return role_context.get_by_role("toolbar", name=locator)
            if normalized == "PRESENTATION":
                return role_context.get_by_role("presentation", name=locator)
            if normalized == "FIGURE":
                return role_context.get_by_role("figure", name=locator)
            if normalized == "ID":
                return role_context.locator(f"#{locator}")
            if normalized == "NAME":
                return role_context.locator(f"[name='{locator}']")
            if normalized == "CLASS":
                return role_context.locator(f".{locator}")

        if normalized == "TEXT":
            return role_context.get_by_text(locator)
        if normalized == "ROLE":
            return role_context.get_by_role(locator.upper())
        if normalized == "ALTTEXT":
            return role_context.get_by_alt_text(locator)
        if normalized == "TITLE":
            return role_context.get_by_title(locator)
        if normalized == "PLACEHOLDER":
            return role_context.get_by_placeholder(locator)
        if normalized == "LABEL":
            return role_context.get_by_label(locator)
        if normalized == "TESTID":
            return role_context.get_by_test_id(locator)

        if chained:
            raise ValueError(f"Unknown chained locator type: {locator_type}")
        raise ValueError(f"Unknown locator type: {locator_type}")
